File: backend/src/main/java/com/example/resourceapp/model/Ticket.java
package com.example.resourceapp.model;

import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import java.time.LocalDateTime;
import java.util.List;
import java.util.ArrayList;

@Document(collection = "tickets")
public class Ticket {

    @Id
    private String id;

    @NotBlank(message = "Resource ID is required")
    private String resourceId;

    @NotBlank(message = "Location is required")
    private String location;

    @NotBlank(message = "Creator ID is required")
    private String createdByUserId;

    @NotBlank(message = "Creator name is required")
    private String createdByUserName;

    @NotNull(message = "Category is required")
    private TicketCategory category;

    @NotBlank(message = "Description is required")
    private String description;

    @NotNull(message = "Priority is required")
    private TicketPriority priority;

    @NotBlank(message = "Preferred contact is required")
    private String preferredContactDetails;

    @NotNull(message = "Status is required")
    private TicketStatus status;

    private String assignedToUserId;
    private String assignedToUserName;

    private String rejectionReason;
    private String resolutionNotes;

    private List<ImageAttachment> imageAttachments;
    private List<TicketComment> comments;

    private LocalDateTime createdAt;
    private LocalDateTime updatedAt;
    private LocalDateTime resolvedAt;
    private LocalDateTime closedAt;

    // Constructors
    public Ticket() {
        this.imageAttachments = new ArrayList<>();
        this.comments = new ArrayList<>();
        this.status = TicketStatus.OPEN;
        this.createdAt = LocalDateTime.now();
        this.updatedAt = LocalDateTime.now();
    }

    // Getters and Setters
    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getResourceId() {
        return resourceId;
    }

    public void setResourceId(String resourceId) {
        this.resourceId = resourceId;
    }

    public String getLocation() {
        return location;
    }

    public void setLocation(String location) {
        this.location = location;
    }

    public String getCreatedByUserId() {
        return createdByUserId;
    }

    public void setCreatedByUserId(String createdByUserId) {
        this.createdByUserId = createdByUserId;
    }

    public String getCreatedByUserName() {
        return createdByUserName;
    }

    public void setCreatedByUserName(String createdByUserName) {
        this.createdByUserName = createdByUserName;
    }

    public TicketCategory getCategory() {
        return category;
    }

    public void setCategory(TicketCategory category) {
        this.category = category;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public TicketPriority getPriority() {
        return priority;
    }

    public void setPriority(TicketPriority priority) {
        this.priority = priority;
    }

    public String getPreferredContactDetails() {
        return preferredContactDetails;
    }

    public void setPreferredContactDetails(String preferredContactDetails) {
        this.preferredContactDetails = preferredContactDetails;
    }

    public TicketStatus getStatus() {
        return status;
    }

    public void setStatus(TicketStatus status) {
        this.status = status;
    }

    public String getAssignedToUserId() {
        return assignedToUserId;
    }

    public void setAssignedToUserId(String assignedToUserId) {
        this.assignedToUserId = assignedToUserId;
    }

    public String getAssignedToUserName() {
        return assignedToUserName;
    }

    public void setAssignedToUserName(String assignedToUserName) {
        this.assignedToUserName = assignedToUserName;
    }

    public String getRejectionReason() {
        return rejectionReason;
    }

    public void setRejectionReason(String rejectionReason) {
        this.rejectionReason = rejectionReason;
    }

    public String getResolutionNotes() {
        return resolutionNotes;
    }

    public void setResolutionNotes(String resolutionNotes) {
        this.resolutionNotes = resolutionNotes;
    }

    public List<ImageAttachment> getImageAttachments() {
        return imageAttachments;
    }

    public void setImageAttachments(List<ImageAttachment> imageAttachments) {
        this.imageAttachments = imageAttachments;
    }

    public List<TicketComment> getComments() {
        return comments;
    }

    public void setComments(List<TicketComment> comments) {
        this.comments = comments;
    }

    public LocalDateTime getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(LocalDateTime createdAt) {
        this.createdAt = createdAt;
    }

    public LocalDateTime getUpdatedAt() {
        return updatedAt;
    }

    public void setUpdatedAt(LocalDateTime updatedAt) {
        this.updatedAt = updatedAt;
    }

    public LocalDateTime getResolvedAt() {
        return resolvedAt;
    }

    public void setResolvedAt(LocalDateTime resolvedAt) {
        this.resolvedAt = resolvedAt;
    }

    public LocalDateTime getClosedAt() {
        return closedAt;
    }

    public void setClosedAt(LocalDateTime closedAt) {
        this.closedAt = closedAt;
    }

    // Helper methods
    public void addImageAttachment(ImageAttachment attachment) {
        if (this.imageAttachments == null) {
            this.imageAttachments = new ArrayList<>();
        }
        if (this.imageAttachments.size() < 3) {
            this.imageAttachments.add(attachment);
        }
    }

    public void addComment(TicketComment comment) {
        if (this.comments == null) {
            this.comments = new ArrayList<>();
        }
        this.comments.add(comment);
    }
}

File: frontend/src/types/Ticket.ts
export type TicketStatus = 'OPEN' | 'IN_PROGRESS' | 'RESOLVED' | 'CLOSED' | 'REJECTED';
export type TicketPriority = 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';
export type TicketCategory =
  | 'EQUIPMENT_MALFUNCTION'
  | 'DAMAGE'
  | 'CLEANING'
  | 'SAFETY_HAZARD'
  | 'CONNECTIVITY_ISSUE'
  | 'MAINTENANCE'
  | 'OTHER';

export interface ImageAttachment {
  id: string;
  fileName: string;
  fileType: string;
  fileData: string; // Base64 encoded
  fileSize: number;
  uploadedAt: string;
}

export interface TicketComment {
  id: string;
  userId: string;
  userName: string;
  content: string;
  createdAt: string;
  updatedAt: string;
}

export interface Ticket {
  id: string;
  resourceId: string;
  location: string;
  createdByUserId: string;
  createdByUserName: string;
  category: TicketCategory;
  description: string;
  priority: TicketPriority;
  preferredContactDetails: string;
  status: TicketStatus;
  assignedToUserId?: string;
  assignedToUserName?: string;
  rejectionReason?: string;
  resolutionNotes?: string;
  imageAttachments: ImageAttachment[];
  comments: TicketComment[];
  createdAt: string;
  updatedAt: string;
  resolvedAt?: string;
  closedAt?: string;
}

export interface CreateTicketRequest {
  resourceId: string;
  location: string;
  category: TicketCategory;
  description: string;
  priority: TicketPriority;
  preferredContactDetails: string;
  imageDataList?: string[]; // Base64 encoded images
}

export interface AddCommentRequest {
  content: string;
}

export interface UpdateTicketStatusRequest {
  status: TicketStatus;
  resolutionNotes?: string;
  rejectionReason?: string;
}

export interface AssignTicketRequest {
  technicianId: string;
  technicianName: string;
}
